File: backend/app/schemas/payment.py
# backend/app/schemas/payment.py
from datetime import datetime
from decimal import Decimal
from enum import Enum
from typing import Any, Dict, List, Optional

from pydantic import UUID4, BaseModel, Field


class PaymentProvider(str, Enum):
    """Поддерживаемые платежные системы"""

    YOOKASSA = "yookassa"
    CASH = "cash"
    # Другие платежные системы могут быть добавлены здесь


class PaymentStatus(str, Enum):
    """Статусы платежей"""

    PENDING = "pending"
    WAITING_FOR_CAPTURE = "waiting_for_capture"
    SUCCEEDED = "succeeded"
    CANCELED = "canceled"
    REFUNDED = "refunded"
    FAILED = "failed"

    @classmethod
    def get_all_statuses(cls) -> List[str]:
        """Получение всех возможных статусов"""
        return [status.value for status in cls]


class SPaymentBase(BaseModel):
    """Базовая схема платежа"""

    provider: str = Field(..., description="Платежная система")
    amount: Decimal = Field(..., description="Сумма платежа")
    currency: str = Field("RUB", description="Валюта платежа")
    payment_method: Optional[str] = Field(None, description="Метод оплаты")


class SPaymentCreate(SPaymentBase):
    """Схема для создания платежа"""

    order_id: UUID4 = Field(..., description="ID заказа")
    return_url: Optional[str] = Field(None, description="URL возврата после оплаты")


class SPayment(SPaymentBase):
    """Схема для отображения платежа"""

    id: UUID4 = Field(..., description="ID платежа")
    order_id: UUID4 = Field(..., description="ID заказа")
    status: str = Field(..., description="Статус платежа")
    provider_payment_id: Optional[str] = Field(
        None, description="ID платежа в системе провайдера"
    )
    confirmation_url: Optional[str] = Field(
        None, description="URL для подтверждения оплаты"
    )
    payment_data: Optional[Dict[str, Any]] = Field(
        None, description="Дополнительные данные платежа"
    )
    created_at: datetime = Field(..., description="Дата создания")
    updated_at: datetime = Field(..., description="Дата обновления")
    paid_at: Optional[datetime] = Field(None, description="Дата оплаты")
    refunded_at: Optional[datetime] = Field(None, description="Дата возврата")

    class Config:
        from_attributes = True


class SPaymentUpdate(BaseModel):
    """Схема для обновления платежа"""

    status: Optional[str] = Field(None, description="Статус платежа")
    provider_payment_id: Optional[str] = Field(
        None, description="ID платежа в системе провайдера"
    )
    payment_data: Optional[Dict[str, Any]] = Field(
        None, description="Дополнительные данные платежа"
    )
    confirmation_url: Optional[str] = Field(
        None, description="URL для подтверждения оплаты"
    )


class SYookassaWebhookRequest(BaseModel):
    """Схема для получения вебхука от ЮКассы"""

    event: str = Field(..., description="Тип события")
    object: Dict[str, Any] = Field(..., description="Данные о платеже")


class SPaymentResponse(BaseModel):
    """Схема ответа для платежа"""

    payment_id: UUID4 = Field(..., description="ID платежа")
    confirmation_url: Optional[str] = Field(None, description="URL для оплаты")
    status: str = Field(..., description="Статус платежа")


class SWidgetPaymentResponse(SPaymentResponse):
    """Схема ответа для платежа с виджетом"""

    confirmation_token: str = Field(..., description="Токен для инициализации виджета")

File: frontend/app/src/types/payment.ts
import { PAYMENT_STATUSES } from "@/constants/payment";
import type { UUID } from "./common";

/**
 * Интерфейс ответа на создание платежа
 */
export interface IPaymentResponse {
  payment_id: UUID;
  confirmation_url: string;
  status: string;
}

/**
 * Интерфейс статуса платежа
 */
export interface IPaymentStatus {
  id: UUID;
  order_id: UUID;
  provider: string;
  provider_payment_id: string;
  status: PaymentStatusType;
  amount: string;
  currency: string;
  payment_method: string;
  created_at: string;
  updated_at: string;
  paid_at: string | null;
}

/**
 * Возможные статусы платежа
 */
export type PaymentStatusType =
  | typeof PAYMENT_STATUSES.PENDING
  | typeof PAYMENT_STATUSES.WAITING_FOR_CAPTURE
  | typeof PAYMENT_STATUSES.SUCCEEDED
  | typeof PAYMENT_STATUSES.CANCELED
  | typeof PAYMENT_STATUSES.REFUNDED
  | typeof PAYMENT_STATUSES.FAILED;

/**
 * Интерфейс результата платежа с дополнительной информацией
 */
export interface IPaymentResult {
  status: PaymentStatusType;
  orderId: UUID;
  paymentId: UUID;
  isSuccessful: boolean;
  error?: string;
}

// Структура для виджета
export interface IWidgetPaymentResponse extends IPaymentResponse {
  confirmation_token: string;
}

// Параметры для настройки виджета
export interface IWidgetOptions {
  confirmation_token: string;
  return_url: string;
  customization?: {
    modal?: boolean;
    payment_methods?: string[];
    colors?: {
      control_primary?: string;
      background?: string;
    };
  };
  error_callback: (error: unknown) => void;
}

export interface YooMoneyCheckoutWidget {
  new (options: {
    confirmation_token: string;
    return_url: string;
    customization?: {
      modal?: boolean;
      payment_methods?: string[];
    };
    error_callback: (error: unknown) => void;
  }): {
    render(containerId?: string): Promise<void>;
  };
}

declare global {
  interface Window {
    YooMoneyCheckoutWidget?: YooMoneyCheckoutWidget;
  }
}
